text_machina/src/extractors/sentence_rewriting.py

```diff
@@ -0,0 +1,67 @@
+from math import ceil
+from random import randint, shuffle, uniform
+from typing import Dict, List
+
+from datasets import Dataset
+
+from ..config import InputConfig
+from ..types import TaskType
+from .base import Extractor
+from .utils import spacy_pipeline
+
+
+class SentenceRewriting(Extractor):
+    """
+    Extractor that fills the prompt template with a sentence that
+    has to be rewritten by an LLM.
+
+    This extractor needs two template placeholders:
+
+
+    This extractor needs two template placeholders:
+        - {sentence}: will be filled with sentence to be rewritten.
+
+    This extractor allows to pass the following arguments in the
+    `extractor_args` field from the config:
+        - percentage_range (List[float]): range delimiting the percentage
+            of sentences to be rewritten. At least one sentence will be
+            always rewritten.
+    """
+
+    def __init__(self, input_config: InputConfig, task_type: TaskType):
+        super().__init__(input_config, task_type)
+        self.args = self.input_config.extractor_args.get(
+            "sentence_rewriting", {}
+        )
+        self.workspace = {"positions": [], "human_spans": []}
+
+    def prepare_human(self, human_texts: List[str]) -> List[str]:
+        return [
+            "".join(doc_sentences)
+            for doc_sentences in self.workspace["human_spans"]
+        ]
+
+    def _extract(self, dataset: Dataset) -> Dict[str, List[str]]:
+        text_column = self.input_config.dataset_text_column
+        texts = spacy_pipeline(
+            texts=dataset[text_column], language=self.input_config.language
+        )
+        sentences_to_rewrite = []
+        for text in texts:
+            sentences = list([sent.text_with_ws for sent in text.sents])
+            self.workspace["human_spans"].append(sentences)
+            n_sentences_to_select = randint(
+                1,
+                ceil(uniform(*self.args["percentage_range"]) * len(sentences)),
+            )
+            sampled_positions = [_ for _ in range(len(sentences))]
+            shuffle(sampled_positions)
+            sampled_positions = sorted(
+                sampled_positions[:n_sentences_to_select]
+            )
```

Comment:
I thought that maybe shuffle was slightly inefficient compared to (random.)sample, so I checked:
```
def foo_a(n=100, m=10):
    sampled_positions = [_ for _ in range(n)]
    shuffle(sampled_positions)
    r =  sorted(sampled_positions[:m])
    return r
def foo_b(n=100, m=10):
    sampled_positions = list(range(n))
    r = sorted([sampled_positions[position] for position in sample(sampled_positions, m)])
    return r

# %timeit foo_a()
# 16.1 µs ± 22.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
# In [26]: %timeit foo_b()
# 3.41 µs ± 20.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Reply:
good point! I don't know why I thought `random.sample` will sample with replacement and avoided using it. Changed now to `random.sample`

Reply:
I think you forgot `sorted` when you changed it.

Reply:
totally. Thanks!